### ISSUE TITLE
Custom script for proxy host.

### ISSUE THREAD

**Issue by zaywalker** [jc21/nginx-proxy-manager#15]
First of all, thanks for the awesome proxy manager! it's very easy to use like synology reverse proxy.

It's almost no problem, but some server won't work properly, so I digged in and I found nginx configuration.

https://portainer.readthedocs.io/en/stable/faq.html?highlight=proxy

```
upstream portainer {
    server ADDRESS:PORT;
}

server {
  listen 80;

  location /portainer/ {
      proxy_http_version 1.1;
      proxy_set_header Connection "";
      proxy_pass http://portainer/;
  }
  location /portainer/api/websocket/ {
      proxy_set_header Upgrade $http_upgrade;
      proxy_set_header Connection "upgrade";
      proxy_http_version 1.1;
      proxy_pass http://portainer/api/websocket/;
  }
}
```

But the problem is I have no idea where to put this in.. I've tried to use the advanced tab's custom configuration, but it doesn't work..

Any help and how to guides would be great!

Sorry about this noob question.;;

**Comment by jc21** [comment]
You're almost there, the advanced config actually gets inserted into the server block and the `location /` is already forwarded, so all you need is the websocket part, this should work:

```
location /api/websocket/ {
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection "upgrade";
    proxy_http_version 1.1;
    proxy_pass http://IPADDRESS:PORT/api/websocket/;
}
```

Of course you will need to replace `IPADDRESS` and `PORT`

**Comment by SGStino** [comment]
can't seem to get this working, is there a way we can debug the generated nginx configs?

the files (2.conf, 3.conf, ...) in /data in the container just include 
```
  location / {
 
    # Proxy!
    include conf.d/include/proxy.conf;
  }
}
```

but the custom config added in the gui isn't there.

**Comment by jc21** [comment]
@SGStino  So this is a bit of a limitation at the moment, you can't specify any custom configuration for hosts that are meant to sit outside of the `server` block. In your case, `map` does in fact sit outside of the block and as a result, the `$connection_upgrade` variable would be available to all nginx configurations options if it were to exist. I haven't found a nice, simple way to illustrate this in the interface yet. The goal of the software is to make life easy for people as much as possible. But I do want to have some more advanced configuration in there for users like yourself.

@zaywalker I'm using the latest Portainer myself (1.16.5 stable Docker version) and I'm not seeing any websocket upgrade requests at all in the interface. Even so, I've added the custom code I first mentioned to my proxy hosts and not only does the config show up on disk as part of the nginx config, it works just fine and reports online.

If you have a look at the [proxy config template file](https://github.com/jc21/nginx-proxy-manager/blob/master/src/backend/templates/proxy_host.conf) you'll see where the `advanced_config` is placed. You'll also see that there is already a `location / {` section which **cannot exist twice** in the same `server` block.

If v1.19 of Portainer absolutely needs the websocket upgrade, you should only need this advanced config content:

```
location /api/websocket/ {
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection "upgrade";
    proxy_http_version 1.1;
    proxy_pass http://$server:$port/api/websocket/;
}
```

When you save the Host in the interface and the host shows "Offline", hover the "Offline" text and you should see the error in the Tooltip. Let me know what that says

**Comment by jc21** [comment]
@zaywalker Ok I was a little bit mistaken, I found out where the api/websocket requests are happening and it was indeed failing for me too. I found the source of the problem and it's actually Portainers documentation that was wrong. This is the advanced config I'm using and I can get to my consoles now:

```
location /api/websocket/ {
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection "upgrade";
    proxy_http_version 1.1;
    proxy_pass http://$server:$port;
}
```
Note that the `proxy_pass` doesn't need a path and if one is set, it ignored the path of the request when it's forwarded. In this specific case, `/api/websocket/exec` was being sent to `/api/websocket/` and returning 400 errors for me.
